src/uct/tcp/tcp_iface.c

```diff
@@ -472,24 +475,57 @@ static UCS_CLASS_INIT_FUNC(uct_tcp_iface_t, uct_md_h md, uct_worker_h worker,
     return status;
 }
 
-static UCS_CLASS_CLEANUP_FUNC(uct_tcp_iface_t)
+static void uct_tcp_iface_ep_list_cleanup(uct_tcp_iface_t *iface,
+                                          ucs_list_link_t *ep_list)
 {
     uct_tcp_ep_t *ep, *tmp;
+
+    ucs_list_for_each_safe(ep, tmp, ep_list, list) {
+        uct_tcp_ep_destroy_internal(&ep->super.super);
+    }
+}
+
+static void uct_tcp_iface_eps_cleanup(uct_tcp_iface_t *iface)
+{
+    ucs_list_link_t *ep_list;
+    khint_t iter;
+
+    uct_tcp_iface_ep_list_cleanup(iface, &iface->ep_list);
+
+    /* don't use kh_foreach* here, because it is not safe */
+    iter = kh_begin(&iface->ep_cm_map);
+    while (iter != kh_end(&iface->ep_cm_map)) {
+        if (!kh_exist(&iface->ep_cm_map, iter)) {
+            iter++;
+            continue;
+        }
+
+        ep_list = kh_value(&iface->ep_cm_map, iter);
+        uct_tcp_iface_ep_list_cleanup(iface, ep_list);
+
+        /* ensure the next iteration */
+        iter = kh_begin(&iface->ep_cm_map);
```

Comment:
not sure it's efficient to start from hash begin every time. any chance it could be O(n)?
why is it not safe to use kh_foreach? as long as we would not delete **keys** from the hash, we should be OK

Reply:
now this is 0(n)
added `uct_tcp_cm_purge` that removes EP from CM hash's list, inserts to iface's EP list and removes all ctx caps